File: CommonUtils/src/main/java/alexey/tools/common/misc/MiscUtils.java
package alexey.tools.common.misc;

import alexey.tools.common.collections.ObjectList;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import java.io.Closeable;
import java.io.IOException;
import java.util.Collections;
import java.util.List;

public class MiscUtils {

    @NotNull
    public static List<IOException> closeAll(@NotNull Iterable<? extends Closeable> closeables) {
        ObjectList<IOException> errors = null;
        for (Closeable closeable : closeables) try {
            closeable.close();
        } catch (IOException error) {
            if (errors == null) errors = new ObjectList<>(4);
            errors.add(error);
        }
        return errors == null ? Collections.emptyList() : errors;
    }



    @NotNull
    @Contract(value = " -> new", pure = true)
    public static Object newObject() {
        return new Object();
    }
}

File: CommonUtils/src/main/kotlin/alexey/tools/common/misc/MiscUtils.kt
package alexey.tools.common.misc

import alexey.tools.common.events.EventBus
import alexey.tools.common.mods.ModLoader
import java.awt.image.BufferedImage
import java.io.Closeable
import java.io.File
import javax.imageio.ImageIO

fun Iterable<Closeable>.close() = tryForEach { it.close() }

fun Array<out Closeable>.close() = tryForEach { it.close() }

fun EventBus.register(modLoader: ModLoader,
                      modsClassesName: String = "Mod",
                      initializeMethodName: String = "initialize") {

    val modEntries = modLoader.findObjects(modsClassesName, Any::class.java)
    for (modEntry in modEntries) silentTry {
        modEntry.javaClass
            .getDeclaredMethod(initializeMethodName, EventBus::class.java)
            .invoke(modEntry, this)
    }
}

fun cropImage(input: String, output: String,
              rows: Int, cols: Int,
              horizontalDelta: Int, verticalDelta: Int = horizontalDelta) {

    val inputImage = ImageIO.read(File(input))

    val partWidth = inputImage.width / cols
    val partHeight = inputImage.height / rows

    val newPartWidth = partWidth - horizontalDelta * 2
    val newPartHeight = partHeight - verticalDelta * 2

    val outputImage = BufferedImage(
        newPartWidth * cols,
        newPartHeight * rows,
        BufferedImage.TYPE_INT_ARGB)

    val g = outputImage.createGraphics()
    try {
        for (y in 0 ..< rows) for (x in 0 ..< cols) {
            val dx = x * newPartWidth
            val dy = y * newPartHeight

            val sx = x * partWidth + horizontalDelta
            val sy = y * partHeight + verticalDelta

            g.drawImage(inputImage,
                dx, dy, dx + newPartWidth, dy + newPartHeight,
                sx, sy, sx + newPartWidth, sy + newPartHeight, null)
        }
    } finally {
        g.dispose()
    }
    ImageIO.write(outputImage, output.getExtensionText(), File(output))
}